sui-execution/latest/sui-move-natives/src/eth_state_proof.rs

```diff
@@ -47,7 +53,67 @@ pub fn verify_message_proof(
     _ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
-    todo!();
+    // Load the cost parameters from the protocol config
+    let verify_message_proof_cost_params = &context
+        .extensions()
+        .get::<NativesCostTable>()
+        .eth_state_proof
+        .clone();
+
+    // Charge the base cost for this operation.
+    native_charge_gas_early_exit!(
+        context,
+        verify_message_proof_cost_params.verify_message_proof_cost_base
+    );
+
+    let cost = context.gas_used();
+
+    let (contract_address, state_root, map_slot, dwallet_id, message, proof) = (
+        pop_arg!(args, Vector).to_vec_u8()?,
+        pop_arg!(args, Vector).to_vec_u8()?,
+        pop_arg!(args, u64),
+        pop_arg!(args, Vector).to_vec_u8()?,
+        pop_arg!(args, Vector).to_vec_u8()?,
+        pop_arg!(args, Vector).to_vec_u8()?,
+    );
+
+    let proof = bcs::from_bytes::<EIP1186ProofResponse>(proof.as_slice())
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let message = String::from_utf8(message.clone())
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let message_map_index = get_message_storage_slot(message.clone(), dwallet_id.clone(), map_slot)
+        .map_err(|_| PartialVMError::new(StatusCode::UNKNOWN_STATUS))?;
+
+    let contract_address = Address::from_slice(contract_address.as_slice());
+    let state_root = Bytes32::try_from(state_root.as_slice())
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let account_proof = create_account_proof(&contract_address, &state_root, &proof);
```

Comment:
I don't like this. Why is this needed?

Reply:
I mean, why do you need to put everything into a struct? this function should just compute the path